Fixes race condition getting IPv4 TCP table on Windows.

AndreLouisCaron opened the pull request:
We've been getting rather frequent cases where the `psutil.Process().connections()` returns an empty list on Windows when we know for a fact that there are open TCP connections on the machine.

We traced down the problem to the fact that there is a race condition in the `GetExtendedTCPTable()` API: since we call it twice (once to get the size, once to read the data), it happends that new connections open between the two calls to the `GetExtendedTCPTable()` function and the table size is too small on the 2nd call.

The solution to this is to call the function in a loop and repeat until we manage to successfully read the data.

I couldn't run the tests on my machine, so I'm sending this as a preliminary patch.   I think the other calls (e.g. for IPv6, and probably UDP) also need this too.  Maybe some refactoring could help move this logic in common somewhere.

giampaolo commented:
This looks reasonable. 
Given that this may loop different times I think it makes sense to release the GIL. You can do so by using `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS` around `GetExtendedTCPTable`.
Also you should do the same for `GetExtendedUDPTable`.
Also please update HISTORY and CREDITS file.

AndreLouisCaron commented:
OK, thanks, will update shortly :-)

AndreLouisCaron commented:
OK.  I updated the code to affect TCP IPv6, UDP IPv4 and UDP Ipv6 too.  I also updated `HISTORY.rst` and `CREDITS`.

Tests are not passing now, still needs a bit of work.

AndreLouisCaron commented:
Added the code to release the GIL, as requested :-)

AndreLouisCaron commented:
@giampaolo I think I addressed everything you asked.  Let me know if this PR needs more work :-)

giampaolo commented:
Sorry I'm busy and still didn't have time to look at this but I promise I will (likely during the week end)

giampaolo commented:
Merged. Thanks.

AndreLouisCaron commented:
Thanks for merging this!

giampaolo commented:
Thanks for the PR. ;)
I will make a release soon.
